### PR TITLE
Upgrade from python 3.9 to 3.11 in hub and singleuser-sample for performance

### PR DESCRIPTION
**Pull request by yuvipanda**
Let's get some of those perf benefits!

https://docs.python.org/3.11/whatsnew/3.11.html and https://docs.python.org/3.11/whatsnew/3.11.html#whatsnew311-faster-cpython

### CODE REVIEW

**Comment by consideRatio** [comment]
This build gave us...

```
#19 17.94 Collecting psycopg2-binary==2.9.4
#19 17.96   Downloading psycopg2-binary-2.9.4.tar.gz (384 kB)
```

There is py311 builds out now for https://pypi.org/project/psycopg2-binary/2.9.5/#files, but also earlier, so I'm not sure why we got the old version. But, anyhow, we should be able to handle the failure related to use of a postgresql database soon I think by upstream work.

**Comment by consideRatio** [comment]
@yuvipanda @manics this now works! My main concern for merging is that it is semi-breaking to upgrade the hub pod's python version.

With the `hub-slim` image pending release, we should go for at least 2.1.0, but should it also be 3.0.0 when changing the python version in the hub image or not? I'm leaning towards 2.1.0 atm.

**Comment by manics** [comment]
@mathbunnyru made the suggestion of waiting for the first patch release of a new Python version in https://github.com/jupyter/docker-stacks/pull/1812#issuecomment-1291578628

**Comment by manics** [comment]
I lean towards a major bump when one of Z2JH's dependencies undergoes a significant change that impacts users or admins. There's also a couple of breaking KubeSpawner PRs that could go in https://github.com/jupyterhub/kubespawner/pulls?q=is%3Aopen+is%3Apr+label%3Abreaking

**Comment by bjornjorgensen** [comment]
This PR is also upgrading to ubuntu 22.04. 
Ubuntu 22.10 is out.

**Comment by consideRatio** [comment]
> This PR is also upgrading to ubuntu 22.04.
Ubuntu 22.10 is out.

The hub image / singleuser-sample image will keep using debian based images (via python:3.11-slim-bullseye), but what you observe in this PR is a change in the CI system we use, where the testing logic is executed from a Ubuntu 22.04 environment instead of a Ubuntu 20.04 environment. What we choose there won't affect the package helm chart etc.

Note that 20.04 and 22.04 are special "long term support" releases compared to 22.10 which isn't, so that is a reason why you may see those stay around even though 22.10, 23.04, 23.10 arrives later - until 24.04 comes which will be a LTS as well.

File: .github/workflows/test-chart.yaml [hunk 1]
```diff
@@ -28,12 +28,12 @@ on:
 
 jobs:
   lint_shell_scripts:
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-22.04
```
[the remark this answers is not in the document]
Comment (consideRatio) [reply .github/workflows/test-chart.yaml:31]: Certainly reasonable, but for a every-two-year update, I'd favor being explicit for ease of comprehension over the reduced maintenance needs to bump it.

**Comment by yuvipanda** [comment]
I feel ok merging given @consideRatio rewrote most of the PR to actually work hehe :D 

Thanks!

**Pull request merged**